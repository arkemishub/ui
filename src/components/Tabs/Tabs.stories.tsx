/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ITabsProps } from "./index";
import React from "react";
import Tabs from "./Tabs";

export default {
  title: "Tabs",
  component: Tabs,
};

export const Uncontrolled = () => {
  return (
    <Tabs>
      <Tabs.Tab>Tab 1</Tabs.Tab>
      <Tabs.Tab>Tab 2</Tabs.Tab>

      <Tabs.TabPanel>Content 1</Tabs.TabPanel>
      <Tabs.TabPanel>Content 2</Tabs.TabPanel>
    </Tabs>
  );
};

export const Controlled = (args: Partial<ITabsProps>) => {
  return (
    <Tabs {...args}>
      <Tabs.Tab>Tab 1</Tabs.Tab>
      <Tabs.Tab>Tab 2</Tabs.Tab>

      <Tabs.TabPanel>Content 1</Tabs.TabPanel>
      <Tabs.TabPanel>Content 2</Tabs.TabPanel>
    </Tabs>
  );
};
