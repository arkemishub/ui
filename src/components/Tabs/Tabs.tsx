/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Tab as HeadlessTab } from "@headlessui/react";
import { useCallback, useEffect, useMemo, useState } from "react";
import type { ITabsProps } from "./Tabs.types";
import { twMerge } from "tailwind-merge";
import Tab from "./Tab";
import TabPanel from "./TabPanel";

function Tabs({ children, onChange, active = 0, className }: ITabsProps) {
  const [selectedTab, setSelectedTab] = useState(active);

  const { tabs, tabPanels, others } = useMemo(() => {
    const rawChildren = Array.isArray(children) ? children : [children];
    return rawChildren.reduce(
      (acc, child) => {
        const childType = child.type;
        if (childType === Tab) {
          acc.tabs.push(child);
        } else if (childType === TabPanel) {
          acc.tabPanels.push(child);
        } else {
          acc.others.push(child);
        }

        return acc;
      },
      {
        tabs: [],
        tabPanels: [],
        others: [],
      }
    );
  }, [children]);

  const handleTabChange = useCallback(
    (index: number) => {
      onChange ? onChange(index) : setSelectedTab(index);
    },
    [onChange]
  );

  useEffect(() => {
    if (typeof active !== "undefined") setSelectedTab(active);
  }, [active]);

  return (
    <HeadlessTab.Group selectedIndex={selectedTab} onChange={handleTabChange}>
      <HeadlessTab.List className={twMerge("tabs__list", className)}>
        {tabs}
      </HeadlessTab.List>
      {others}
      <HeadlessTab.Panels>{tabPanels}</HeadlessTab.Panels>
    </HeadlessTab.Group>
  );
}

Tabs.Tab = Tab;
Tabs.TabPanel = TabPanel;

export default Tabs;
