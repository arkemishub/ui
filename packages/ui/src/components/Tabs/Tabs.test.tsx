/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Tabs from "./Tabs";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("Tabs", () => {
  test("should match snapshot", () => {
    const { asFragment } = render(
      <Tabs active={0}>
        <Tabs.Tab>Tab 1 </Tabs.Tab>
        <Tabs.Tab>Tab 2 </Tabs.Tab>
        <Tabs.TabPanel> Panel 1 </Tabs.TabPanel>
        <Tabs.TabPanel> Panel 2</Tabs.TabPanel>
      </Tabs>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should show next tab", async () => {
    const { getByText } = render(
      <Tabs active={0}>
        <Tabs.Tab>Tab 1</Tabs.Tab>
        <Tabs.Tab>Tab 2</Tabs.Tab>
        <Tabs.TabPanel>Panel 1</Tabs.TabPanel>
        <Tabs.TabPanel>Panel 2</Tabs.TabPanel>
      </Tabs>
    );

    await userEvent.click(getByText("Tab 2"));

    expect(getByText("Panel 2")).toBeVisible();
  });

  test("should not render not recognized children", () => {
    const { queryAllByText } = render(
      <Tabs active={0}>
        <Tabs.Tab>Tab 1</Tabs.Tab>
        <Tabs.Tab>Tab 2</Tabs.Tab>
        <Tabs.TabPanel>Panel 1</Tabs.TabPanel>
        <Tabs.TabPanel>Panel 2</Tabs.TabPanel>
        <div>Not recognized</div>
      </Tabs>
    );

    expect(queryAllByText("Not recognized")).toHaveLength(0);
  });
});
