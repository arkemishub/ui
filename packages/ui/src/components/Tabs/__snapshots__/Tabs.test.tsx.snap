// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-orientation="horizontal"
    class="tabs__list"
    role="tablist"
  >
    <button
      aria-controls="headlessui-tabs-panel-:r2:"
      aria-selected="true"
      class="tab tab--selected"
      data-headlessui-state="selected"
      id="headlessui-tabs-tab-:r0:"
      role="tab"
      tabindex="0"
      type="button"
    >
      Tab 1 
    </button>
    <button
      aria-controls="headlessui-tabs-panel-:r3:"
      aria-selected="false"
      class="tab"
      data-headlessui-state=""
      id="headlessui-tabs-tab-:r1:"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 2 
    </button>
  </div>
  <div>
    <div
      aria-labelledby="headlessui-tabs-tab-:r0:"
      class="tab__panel"
      data-headlessui-state="selected"
      id="headlessui-tabs-panel-:r2:"
      role="tabpanel"
      tabindex="0"
    >
       Panel 1 
    </div>
    <span
      aria-labelledby="headlessui-tabs-tab-:r1:"
      id="headlessui-tabs-panel-:r3:"
      role="tabpanel"
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
      tabindex="-1"
    />
  </div>
</DocumentFragment>
`;
